### PR TITLE
fixes array merge issues, adds validation + shaped error messages

### PR DESCRIPTION
**Pull request by ikeforrest**

### CODE REVIEW

File: src/withForm/index.js [hunk 1]
```diff
@@ -73,51 +80,82 @@ const processInput = ({ apiSchema, input }) => {
...
+        [schema]: processInput({ apiSchema, input }),
+        [formData]: omit(
+          {
+            ...props[dataKey]
+          },
+          ["__typename"]
+        ),
+        [formErrors]: {
+          errors: null,
+          disabled: true
```
Comment (pckilgore) [inline src/withForm/index.js]: Can we use a different name here, maybe `submitDisabled` or something?  In bigger forms it's going to beg the question 'what is disabled?`.  Right now I *think* this is for the submit button, a better name would self-document a lot better.

File: src/withForm/index.js [hunk 2]
```diff
@@ -73,51 +80,82 @@ const processInput = ({ apiSchema, input }) => {
...
+        [formUpdate]: state => value => ({
           ...state,
-          [`${formName}FormData`]: deepmerge(
-            state[`${formName}FormData`],
-            value
-          )
-        })
+          [formData]: merge(state[formData], value)
+        }),
+        setErrors: () => value => ({ [formErrors]: value })
```
Comment (pckilgore) [inline src/withForm/index.js]: This wipes state except for the `[formErrors]` key/value pair?
```suggestion
        setErrors: state => value => ({ ...state, [formErrors]: value })
```

File: src/withForm/index.js [hunk 3]
```diff
@@ -73,51 +80,82 @@ const processInput = ({ apiSchema, input }) => {
...
+    withPropsOnChange([formData], ({ setErrors, ...props }) =>
+      _.debounce(setErrors(validator(props[formData], props[schema])), 2000)
     )
   );
 };
 
 export default withForm;
+
+const validator = (formData, JSONSchema) => {
+  const result = validate(removeNullKeys(formData), JSONSchema);
```
Comment (pckilgore) [inline src/withForm/index.js]: Do we need to do all this removeNullKeys stuff if the schema now allows the `null` type?

File: src/withForm/index.js [hunk 4]
```diff
@@ -73,51 +80,82 @@ const processInput = ({ apiSchema, input }) => {
...
+  const errors = {};
+
+  result.errors.map(({ argument, property, message, ...rest }) => {
```
Comment (pckilgore) [inline src/withForm/index.js]: If this was re-written inside a single reducer I think we could avoid a lot of the confusion with the immutable methods (map/reduce/slice) and the mutable methods (shift) and mutation of `errors` here.

At the very least, I would appreciate variable names for future reading.

File: src/withForm/index.js [hunk 5]
```diff
@@ -73,51 +80,82 @@ const processInput = ({ apiSchema, input }) => {
...
+      _.debounce(setErrors(validator(props[formData], props[schema])), 2000)
     )
   );
 };
 
 export default withForm;
+
+const validator = (formData, JSONSchema) => {
+  const result = validate(removeNullKeys(formData), JSONSchema);
+  const errors = {};
```
Comment (pckilgore) [inline src/withForm/index.js]:
```suggestion
  let errors = {};
```
Mutated later.

File: src/withForm/index.js [hunk 6]
```diff
@@ -1,9 +1,16 @@
...
 import omit from "omit-deep";
+import { validate } from "jsonschema";
+
 import { mb } from "../utils/vendor/mb.js";
 import { humanTitles } from "../utils/string";
-import * as deepmerge from "deepmerge";
+import merge from "../utils/merge";
 
 const scalars = {
   awsemail: "string",
```
[the remark this answers is not in the document]
Comment (pckilgore) [reply src/withForm/index.js]: (it's below)...I can't comment on code not a part of this PR.

[the remark this answers is not in the document]
Comment (ikeforrest) [reply src/withForm/index.js]: removes the if statement that would happen below

File: src/withForm/index.js [hunk 7]
```diff
@@ -73,51 +80,82 @@ const processInput = ({ apiSchema, input }) => {
...
+    withPropsOnChange([formData], ({ setErrors, ...props }) =>
+      _.debounce(setErrors(validator(props[formData], props[schema])), 2000)
     )
   );
 };
 
 export default withForm;
+
+const validator = (formData, JSONSchema) => {
+  const result = validate(removeNullKeys(formData), JSONSchema);
```
Comment (ikeforrest) [inline src/withForm/index.js]:
```suggestion
  const result = validate(formData, JSONSchema);
```

File: src/withForm/index.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (pckilgore) [reply src/withForm/index.js]: :+1:

**Pull request merged**